src/views/Senate/Documents.js

```diff
@@ -0,0 +1,220 @@
+import InlineEthereumTransaction from 'components/InlineEthereumTransaction';
+import View from 'components/View';
+import BridgeForm from 'form/BridgeForm';
+import convertToInt from 'lib/convertToInt';
+
+function useCastDocumentVote() {
+  const { contracts } = useNetwork();
+  const { pointCursor } = usePointCursor();
+
+  const _contracts = need.contracts(contracts);
+  const _point = need.point(pointCursor);
+
+  return useEthereumTransaction(
+    useCallback(
+      (hash, accept) => {
+        return azimuth.ecliptic.castDocumentVote(
+          _contracts,
+          _point,
+          hash,
+          accept
+        );
+      },
+      [_contracts, _point]
+    ),
+    useCallback(() => null, []), //TODO update polls[hash].hasVoted ?
+    GAS_LIMITS.DEFAULT //TODO update once we have a decent sample size
+  );
+}
+
+function renderDocumentHash(doc) {
+  return doc.slice(0, 34) + '\n  ' + doc.slice(34);
+}
+
+export default function Documents() {
+  const { pop } = useLocalRouter();
+  const { contracts } = useNetwork();
+  const { pointCursor } = usePointCursor();
+
+  const _contracts = need.contracts(contracts);
+  const _point = convertToInt(need.point(pointCursor), 10);
+
+  const { construct, bind } = useCastDocumentVote();
+
+  // const [documentHash, setDocumentHash] = useState('0x...');
+  const [majorities, setMajorities] = useState([]);
+  const [proposals, setProposals] = useState([]);
+  const [polls, setPolls] = useState({});
+  const [loading, setLoading] = useState(true);
+  const [votingOn, setVotingOn] = useState({});
+
+  //TODO update every second?
+  const now = Math.round(new Date().getTime() / 1000);
+
+  useMemo(async () => {
+    let proposals = await azimuth.polls.getDocumentProposals(_contracts);
+    let majorities = await azimuth.polls.getDocumentMajorities(_contracts);
+    let polls = {};
+    for (let doc of proposals) {
+      let poll = await azimuth.polls.getDocumentPoll(_contracts, doc);
+      poll.endTime = convertToInt(poll.start) + convertToInt(poll.duration);
+      poll.hasVoted = await azimuth.polls.hasVotedOnDocumentPoll(
+        _contracts,
+        _point,
+        doc
+      );
+      polls[doc] = poll;
+    }
+    setMajorities(majorities);
+    setProposals(proposals);
+    setPolls(polls);
+    setLoading(false);
+    return;
+  }, [_contracts, _point]);
+
+  // eslint-disable-next-line
+  const [open, closed] = useMemo(() => {
+    let open = [];
+    let closed = [];
+    if (loading) return [open, closed];
+    for (let doc of proposals) {
+      if (now > polls[doc].endTime) {
+        closed.push(doc);
+      } else {
+        open.push(doc);
+      }
+    }
+    return [open, closed];
+  }, [loading, proposals, polls, now]);
+
+  const doVote = useCallback(
+    (hash, accept) => {
+      construct(hash, accept);
+      setVotingOn({ hash, accept });
+    },
+    [construct, setVotingOn]
+  );
+
+  const didVote = useCallback(() => {
+    polls[votingOn].hasVoted = true;
```

Comment:
Does this rerender correctly? I'd be surprised if it did because the object references are the same. It's more idiomatic to use pass a function to setPolls and update the state immutably. Plus, it lets you drop polls from the dependency list

Reply:
It does, actually! (After fixing the other ways in which the logic here was broken, that is...) But I'll update based on your advice regardless. Thanks!